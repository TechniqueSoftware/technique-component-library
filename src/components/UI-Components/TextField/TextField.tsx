import * as React from 'react';
import { default as MuiTextField, TextFieldProps as MuiTextFieldProps } from '@material-ui/core/TextField';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';

function determineLabelWidth(inputLabel: React.RefObject<HTMLLabelElement>) {
  return inputLabel && inputLabel.current && inputLabel.current.offsetWidth;
}

const useStyles = makeStyles({
  visuallyHidden: {
    marginLeft: '-9999px'
  },
  appendedText: {
    marginLeft: '20px',
  }
});

export function TextField(props: TextFieldProps) {

  const inputLabel: React.RefObject<HTMLLabelElement> = React.useRef<HTMLLabelElement>(null);
  const [labelWidth, setLabelWidth] = React.useState<null | number>(0);

  // @ts-ignore
  TextField.defaultProps = {
    variant: 'outlined',
    margin: 'dense'
  };

  const {
    isInvisible,
    hideLabel,
    maxWidth,
    appendedText,
    min,
    max,
    ...propsToPassOn
  } = props;
  const classes = useStyles();

  React.useEffect(() => {
    if (!isInvisible) {
      setLabelWidth(determineLabelWidth(inputLabel));
    }
  }, [isInvisible, propsToPassOn.label]);

  const inputLabelProps = propsToPassOn.InputLabelProps;

  const decoratedProps: TextFieldProps = {
    ...propsToPassOn,
    InputLabelProps: {
      ...inputLabelProps,
      ref: inputLabel,
    },
    inputProps: {
      ...propsToPassOn.inputProps,
      min,
      max
    },
    SelectProps: {
      ...propsToPassOn.SelectProps,
      // @ts-ignore
      style: { maxWidth: (maxWidth ? maxWidth : 'initial'), minWidth: (hideLabel ? 0 : labelWidth) + 48 }
    }, InputProps: {
      ...props.InputProps,
      labelWidth,
    } as any
  };

  if (hideLabel && decoratedProps.InputLabelProps) {
    decoratedProps.InputLabelProps.classes = { root: classes.visuallyHidden };
    decoratedProps.InputLabelProps.shrink = false;
  }

  const muiTextField = <MuiTextField {...decoratedProps} />;

  if (props.appendedText) {
    return (
      <Grid container alignItems="center" onClick={decoratedProps.onClick}>
        {muiTextField}
        <Typography>
          <span className={classes.appendedText}>{appendedText}</span>
        </Typography>
      </Grid>
    );
  }

  return muiTextField;
}

export type TextFieldProps = MuiTextFieldProps & {
  // The label for the TextField relies on element being visible to determine the label width. This prop gives
  // us a way to indicate when the TextField is invisible and changing to visible so that we can recalculate the width.
  isInvisible?: boolean,
  hideLabel?: boolean,
  appendedText?: string,
  maxWidth?: number,
  defaultProps?: any,
  min?: string,
  max?: string
};

export default TextField as React.ComponentType<TextFieldProps>;
