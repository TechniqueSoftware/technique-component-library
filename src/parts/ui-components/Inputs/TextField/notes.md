# TextField
This component wraps Material-UI's [TextField](https://material-ui.com/components/text-fields/#outlined) and applies our defaults.
It also adds the options to hide the label and append text.


## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/CR-664
