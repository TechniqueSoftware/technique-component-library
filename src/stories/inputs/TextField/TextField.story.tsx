import * as React from 'react';
import { createCUICStoriesOf } from '../../utils/utils';
import { MODULE_NAME } from '../constants';
import { clubOsTheme } from '@clubos-ui-components/core';
import { TextField, TextFieldProps } from '@clubos-ui-components/core/src/components/TextField/TextField';
import { boolean, select, text } from '@storybook/addon-knobs';
import ThemeProvider from '@material-ui/styles/ThemeProvider';

type TextFieldStoryProps = {
  label: string;
  variant: 'standard' | 'outlined' | 'filled' | undefined | any;
  disabled: boolean;
  hideLabel: boolean;
  appendedText: string;
};

function TextFieldStory(props: TextFieldStoryProps) {
  const { label, variant, disabled, hideLabel, appendedText } = props;
  const [value, setValue] = React.useState('');
  const textFieldProps: TextFieldProps = {
    label,
    variant,
    disabled,
    hideLabel,
    appendedText,
    value,
    onChange: (event) => setValue(event.target.value),
  };

  return (
    <div>
      <ThemeProvider theme={clubOsTheme}>
        <TextField {...textFieldProps} />
      </ThemeProvider>
    </div>
  );
}

createCUICStoriesOf(MODULE_NAME.INPUTS, module)
  .addCUICStory({
    render: () => {
      const label = text('Label', 'Search');
      const variant = select('variant', ['standard', 'filled', 'outlined'], 'outlined');
      const disabled = boolean('disabled', false);
      const hideLabel = boolean('hide label', false);
      const appendedText = text('append text', '');
      return (<TextFieldStory
        label={label}
        variant={variant}
        disabled={disabled}
        hideLabel={hideLabel}
        appendedText={appendedText}
      />);
    },
    name: 'TextField',
    notes: require('./notes.md').default
  });
