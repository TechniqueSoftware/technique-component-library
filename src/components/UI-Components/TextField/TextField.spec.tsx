import * as React from 'react';
import { TextField } from './TextField';
import { shallow } from 'enzyme';

describe('TextField', () => {
  test('should set variant default to outline', () => {
    const shallowWrapper = shallow(<TextField />);
    expect(shallowWrapper.debug()).toMatchSnapshot();

    const innerComponent = shallowWrapper.find('WithStyles(ForwardRef(TextField))');
    expect(innerComponent.props()).toMatchSnapshot();
  });
});
