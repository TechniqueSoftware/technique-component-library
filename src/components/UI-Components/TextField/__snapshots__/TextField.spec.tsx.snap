// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should set variant default to outline 1`] = `"<WithStyles(ForwardRef(TextField)) InputLabelProps={{...}} inputProps={{...}} SelectProps={{...}} InputProps={{...}} />"`;

exports[`TextField should set variant default to outline 2`] = `
Object {
  "InputLabelProps": Object {
    "ref": Object {
      "current": null,
    },
  },
  "InputProps": Object {
    "labelWidth": 0,
  },
  "SelectProps": Object {
    "style": Object {
      "maxWidth": "initial",
      "minWidth": 48,
    },
  },
  "inputProps": Object {
    "max": undefined,
    "min": undefined,
  },
}
`;
